### DISCUSSION TITLE
Getting dip and azimuth from mesh directly from PyVista

### DISCUSSION THREAD

**Discussion by AlexanderJuestel** [pyvista/pyvista#4071] [Ideas]
Hello, 

I am currently preparing some notebooks for our upcoming GemGIS JOSE publication (https://jose.theoj.org/papers/5e6ad1eedf43a1175c41e188b2881941) and wanted to compare the orientations we provided for the geological modeling in GemPy with the orientations of the resulting PyVista meshes representing geological boundary surfaces. 

It is relatively easy to calculate the normals of the mesh and to calculate the dip angle and dip direction from it (see for example https://courses.eas.ualberta.ca/eas421/formulasheets/formulasheetxythetaP12010.pdf) 

```python
# mesh is extracted geological boundary from GemPy
mesh = mesh.compute_normals()
mesh['Normals'][0]
```
array([ 0.0862411, -0.4506283,  0.8885362], dtype=float32)

```python
dip = np.rad2deg(-np.arccos(mesh['Normals'][0][2]))
dip
```
-27.310118 (pointing downwards)
```python

# adding 270° as the second entry is negative and subtracting 90° to make it align with the axis of GemGIS/GemPy/PyVista
azimuth = np.rad2deg(np.arctan(mesh['Normals'][0][0]/mesh['Normals'][0][1]))+270-90
azimuth 
```
169.16575717926025

Similar to the shown normal vectors, we could implement a function to return the dip and azimuth at the same places. This can further be used for some statistics or so. Maybe something similar is already implemented into PyVista that I have not found yet

![image](https://user-images.githubusercontent.com/45469915/221538437-bd549d6a-2d2b-494b-9af9-87c16056c111.png)

**Comment by banesullivan** [comment]
Nice example! This is exactly how I'd go about this... though maybe decimating the mesh to get more averages

**Comment by AlexanderJuestel** [comment]
What makes the whole thing interesting again is adding the dip and azimuth as arrays to the mesh (which came to last night) and plot them as scalars :D 

```
mesh['dips'] = dips
mesh['azimuths'] = azimuths
```
That way I could see not only applications for structural geological modeling ;)  

![image](https://user-images.githubusercontent.com/45469915/221786619-0fd80139-2f5f-4d23-ab00-7cca19a2b8d7.png)
![image](https://user-images.githubusercontent.com/45469915/221786646-732f1cd4-461b-4554-86d8-3db7a2d375d8.png)
